src/include/txn.i

```diff
@@ -193,6 +193,166 @@ __wt_timestamp_subone(wt_timestamp_t *ts)
+{
+	WT_CURSOR *cursor;
+	WT_DECL_RET;
+	WT_TXN *txn;
+	WT_UPDATE *upd;
+	const char *open_cursor_cfg[] = {
+	    WT_CONFIG_BASE(session, WT_SESSION_open_cursor), NULL };
+
+	txn = &session->txn;
+
+	switch (op->type) {
+	case WT_TXN_OP_NONE:
+		break;
+	case WT_TXN_OP_BASIC_COL:
+	case WT_TXN_OP_BASIC_ROW:
+	case WT_TXN_OP_INMEM_COL:
+	case WT_TXN_OP_INMEM_ROW:
+		WT_RET(__wt_open_cursor(session,
+		    op->btree->dhandle->name, NULL, open_cursor_cfg, &cursor));
+
+		/*
+		 * Transaction prepare is cleared temporarily as cursor
+		 * functions are not allowed for prepared transactions.
+		 */
+		F_CLR(txn, WT_TXN_PREPARE);
+		if (op->type == WT_TXN_OP_BASIC_ROW ||
+		    op->type == WT_TXN_OP_INMEM_ROW)
+			__wt_cursor_set_raw_key(cursor, &op->u.op_row.key);
+		else
+			((WT_CURSOR_BTREE *)cursor)->iface.recno =
+			    op->u.op_col.recno;
+		F_SET(txn, WT_TXN_PREPARE);
+
+		WT_WITH_BTREE(session,
+		    op->btree, ret = __wt_btcur_search_uncommitted(
+		    (WT_CURSOR_BTREE *)cursor, &upd));
+		WT_TRET(cursor->close(cursor));
+		WT_RET(ret);
+		WT_ASSERT(session, upd != NULL);
+		op->u.op_upd = upd;
+
+		for (; upd != NULL; upd = upd->next) {
+			/* Process all the uncommitted updates of this txn. */
+			 if (upd->txnid != txn->id)
+				continue;
+
+			if (!commit) {
+				/* Rollback is just to update txn id. */
+				upd->txnid = WT_TXN_ABORTED;
+				continue;
+			}
+
+			/*
+			 * Newer updates are inserted at head of update chain,
+			 * and txn operations are added at the tail of the txn
+			 * modify chain.
+			 *
+			 * For example, a transaction has modified [k,v] as
+			 * [k,v] -> [k, u1]   (txn_op : txn_op1)
+			 * [k, u1] -> [k, u2] (txn_op : txn_op2)
+			 * update chain : u2->u1
+			 * txn_mod      : txn_op1->txn_op2.
+			 *
+			 * we save only key in the txn op, hence during
+			 * commit/rollback we cannot identify "txn_op1"
+			 * corresponds to either "u2" or "u1".
+			 *
+			 * To make things simpler we will handle all the updates
+			 * that match the key saved in a txn op. As a result,
+			 * multiple updates of a key will be resolved as part of
+			 * the first txn op resolution of that key, and
+			 * subsequent txn op resolution of the same key will be
+			 * effectively a no-op.
+			 *
+			 * In the above example, we will resolve "u2" and "u1"
+			 * as part of resolving "txn_op1" and will not do any
+			 * significant thing as part of "txn_op2".
+			 */
+			if ( upd->prepare_state == WT_PREPARE_RESOLVED)
+				break;
+
+			/*
+			 * In case of a prepared transaction, the order of
+			 * modification of the prepare timestamp to commit
+			 * timestamp in the update chain will not affect the
+			 * data visibility, a reader will encounter a prepared
+			 * update resulting in prepare conflict.
+			 *
+			 * As updating timestamp might not be an atomic
+			 * operation, we will manage using state.
+			 */
+			upd->prepare_state = WT_PREPARE_LOCKED;
```

Comment:
Did you consider setting the update in the WT_TXN_OP structure and using a call to `__wt_txn_op_set_timestamp` here? It'd save duplicating this subtle code.

Reply:
we cannot call `__wt_txn_op_set_timestamp` as this txn will have a commit timestamp, to avoid the duplication of this subtle code, I have added an inline function `__txn_resolve_prepared_update`.